Crash in the PyPI importer?

Hello OSV folks!

**Describe the bug**

I was looking at pypa/advisory-database's OSV ingestion, and I noticed that the OSV ingestor (which is implemented in this repo but runs in the CI there) appears to be failing.

Example failing job:

https://github.com/pypa/advisory-database/actions/runs/23661246588/job/68931794611

Excerpted failure:

```
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
main.loadExisting.func1({0x1786a097e090, 0x23}, {0xfec950?, 0x1786cd6bf860?}, {0x0?, 0x0?})
	/home/runner/go/pkg/mod/github.com/google/osv/vulnfeeds@v0.0.0-20260327050509-02cae864f15f/cmd/pypi/main.go:66 +0x5c5
path/filepath.walk({0x1786a097e090, 0x23}, {0xfec950, 0x1786cd6bf860}, 0x17870c20d590)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.26.1.linux-amd64/src/path/filepath/path.go:346 +0x105
path/filepath.walk({0x178704ad45b0, 0xd}, {0xfec950, 0x1786cd6bf790}, 0x17869e14b590)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.26.1.linux-amd64/src/path/filepath/path.go:370 +0x245
path/filepath.walk({0x7fff33e5fc85, 0x5}, {0xfec950, 0x1786cd6bf5f0}, 0x17869e14b590)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.26.1.linux-amd64/src/path/filepath/path.go:370 +0x245
path/filepath.Walk({0x7fff33e5fc85, 0x5}, 0x17870c20d590)
	/home/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.26.1.linux-amd64/src/path/filepath/path.go:428 +0x66
main.loadExisting({0x7fff33e5fc85, 0x5})
	/home/runner/go/pkg/mod/github.com/google/osv/vulnfeeds@v0.0.0-20260327050509-02cae864f15f/cmd/pypi/main.go:43 +0x4f
main.main()
	/home/runner/go/pkg/mod/github.com/google/osv/vulnfeeds@v0.0.0-20260327050509-02cae864f15f/cmd/pypi/main.go:132 +0x505
Error: Process completed with exit code 2.
```

**To Reproduce**

I don't have a local reproducer, sorry. You can see that it's been consistently failing in CI for a while here:

https://github.com/pypa/advisory-database/actions/workflows/auto_import.yaml

**Expected behaviour**

I expected `github.com/google/osv/vulnfeeds/cmd/pypi@master` to not crash.

**Screenshots**

N/A

**Additional context**

I *suspect* this ingestor failure means that there's a large backlog of OSV records that haven't been assigned PYSEC IDs.

CC @miketheman @sethmlarson

--- a/vulnfeeds/cmd/pypi/main.go
+++ b/vulnfeeds/cmd/pypi/main.go
@@ -64,9 +64,6 @@ func loadExisting(vulnsDir string) (map[string]bool, error) {
 		}
 
 		for _, affected := range vuln.GetAffected() {
-			if affected.GetPackage() == nil {
-				continue
-			}
 			pkgName := affected.GetPackage().GetName()
 			if pkgName == "" {
 				continue

--- a/vulnfeeds/cmd/pypi/main.go
+++ b/vulnfeeds/cmd/pypi/main.go
@@ -63,9 +63,18 @@ func loadExisting(vulnsDir string) (map[string]bool, error) {
 			return fmt.Errorf("failed to parse %s: %w", path, err)
 		}
 
-		ids[vuln.GetId()+"/"+vuln.GetAffected()[0].GetPackage().GetName()] = true
-		for _, alias := range vuln.GetAliases() {
-			ids[alias+"/"+vuln.GetAffected()[0].GetPackage().GetName()] = true
+		for _, affected := range vuln.GetAffected() {
+			if affected.GetPackage() == nil {
+				continue
+			}
+			pkgName := affected.GetPackage().GetName()
+			if pkgName == "" {
+				continue
+			}
+			ids[vuln.GetId()+"/"+pkgName] = true
+			for _, alias := range vuln.GetAliases() {
+				ids[alias+"/"+pkgName] = true
+			}
 		}
 
 		return nil

--- a/vulnfeeds/cmd/pypi/main_test.go
+++ b/vulnfeeds/cmd/pypi/main_test.go
@@ -1,0 +1,99 @@
+// Copyright 2026 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+func TestLoadExisting(t *testing.T) {
+	tmpDir := t.TempDir()
+
+	// 1. Write a valid vulnerability YAML
+	validYaml := `
+vulnerability:
+  id: PYSEC-2021-123
+  affected:
+    - package:
+        name: foo-pkg
+        ecosystem: PyPI
+  aliases:
+    - CVE-2021-12345
+`
+	if err := os.WriteFile(filepath.Join(tmpDir, "valid.yaml"), []byte(validYaml), 0600); err != nil {
+		t.Fatalf("failed to write valid YAML: %v", err)
+	}
+
+	// 2. Write a vulnerability YAML with empty/missing affected block
+	missingAffectedYaml := `
+vulnerability:
+  id: PYSEC-2021-456
+  aliases:
+    - CVE-2021-67890
+`
+	if err := os.WriteFile(filepath.Join(tmpDir, "missing_affected.yaml"), []byte(missingAffectedYaml), 0600); err != nil {
+		t.Fatalf("failed to write YAML with missing affected: %v", err)
+	}
+
+	// 3. Write a vulnerability YAML with affected block but missing package
+	missingPackageYaml := `
+vulnerability:
+  id: PYSEC-2021-789
+  affected:
+    - {}
+`
+	if err := os.WriteFile(filepath.Join(tmpDir, "missing_package.yaml"), []byte(missingPackageYaml), 0600); err != nil {
+		t.Fatalf("failed to write YAML with missing package: %v", err)
+	}
+
+	// Run loadExisting
+	got, err := loadExisting(tmpDir)
+	if err != nil {
+		t.Fatalf("loadExisting failed: %v", err)
+	}
+
+	// Check expected IDs
+	expected := map[string]bool{
+		"PYSEC-2021-123/foo-pkg": true,
+		"CVE-2021-12345/foo-pkg": true,
+	}
+
+	if len(got) != len(expected) {
+		t.Errorf("loadExisting returned map of size %d, want %d", len(got), len(expected))
+	}
+
+	for k, v := range expected {
+		if !got[k] {
+			t.Errorf("Expected map to contain %s", k)
+		}
+		if got[k] != v {
+			t.Errorf("Expected map[%s] to be %t, got %t", k, v, got[k])
+		}
+	}
+
+	// Ensure invalid/skipped entries are NOT present
+	unexpectedKeys := []string{
+		"PYSEC-2021-456/foo-pkg",
+		"CVE-2021-67890/foo-pkg",
+		"PYSEC-2021-789/foo-pkg",
+	}
+	for _, k := range unexpectedKeys {
+		if got[k] {
+			t.Errorf("Map unexpectedly contains %s", k)
+		}
+	}
+}

--- a/vulnfeeds/vulns/vulns.go
+++ b/vulnfeeds/vulns/vulns.go
@@ -769,7 +769,7 @@ func GetCPEs(cpeApplicability []models.CPE, metrics *models.ConversionMetrics) [
 // FromYAML deserializes a Vulnerability from a YAML reader.
 func FromYAML(r io.Reader) (*Vulnerability, error) {
 	decoder := yaml.NewDecoder(r)
-	var vuln Vulnerability
+	vuln := Vulnerability{Vulnerability: &osvschema.Vulnerability{}}
 	err := decoder.Decode(&vuln)
 	if err != nil {
 		return nil, err
@@ -781,7 +781,7 @@ func FromYAML(r io.Reader) (*Vulnerability, error) {
 // FromJSON deserializes a Vulnerability from a JSON reader.
 func FromJSON(r io.Reader) (*Vulnerability, error) {
 	decoder := json.NewDecoder(r)
-	var vuln Vulnerability
+	vuln := Vulnerability{Vulnerability: &osvschema.Vulnerability{}}
 	err := decoder.Decode(&vuln)
 	if err != nil {
 		return nil, err

